### DISCUSSION TITLE
customEditor API

### DISCUSSION THREAD

**Discussion by ManuelHentschel** [James-Yu/LaTeX-Workshop#3951] [Q&A]
Hi, follow-up to #3337 and #3338 here. Is using the customEdior API (without the workaround of reopening each requested document) something you would consider at the moment?

I implemented this in https://github.com/ManuelHentschel/LaTeX-Workshop/tree/customEditor, based on the latest version of this repo, and would be happy to create a PR, work out any remaining bugs, or adapt design choices to your preferences, if you are interested.

**Comment by James-Yu** [comment]
We tried, but reverted in d6f0bad3fd1e5b672da455d91d970f755e0a2d5f due to regressions. The most prominent one is, when the pdf is removed (e.g., during compilation. Some tools first remove then create a new pdf), custom editors will be closed by vscode. This behavior cannot be controlled, making e.g. #3642 #3659

**Reply by ManuelHentschel** [reply]
Thanks for the quick reply!

Whether editors of deleted files are closed by vscode should be controllable with the setting `"workbench.editor.closeOnFileDelete"` which is `false` by default. With this setting I could not reproduce the two linked issues (windows 10, vscode 1.79, Latex-Workshop from right before d6f0bad and https://github.com/ManuelHentschel/LaTeX-Workshop/tree/customEditor).

**Reply by ManuelHentschel** [reply]
In case there are some scenarios where this bug is anavoidable, would you consider having the two approaches in the extension and choosing based on the scenario/user preference? Most of the code between them is the same (as can be seen by the small size of d6f0bad), the key difference seems to be only, who "owns" the underlying webview. A reasonable way to choose between the two could be:

- Pdf viewer launched by a Latex-Workshop command: in a webview managed entirely by the extension
- Pdf viewer launched by opening a pdf file (e.g. clicking on it in the side bar): using the customEditor approach

**Reply by James-Yu** [reply]
Unfortunately, I can still reproduce with 723705e . You may try like this: click the psf from the sidebar, then right click to delete the file. The opened editor will close despite the config is set to false.

**Reply by James-Yu** [reply]
Also, I’m against using two different approaches for viewing from the extension or clicking on the sidebar. This is 1. creating ambiguity if multiple tabs are opened from these two approaches, 2. effectively letting the extension to maintain codes solely for pdf viewing, which is not the objective of the extension, 3. hard to say whether the html listener to viewer events should be enabled or not on sidebar-clicked viewer.

May I bring up the question again, that is, what is missing from the current approach?

**Reply by ManuelHentschel** [reply]
> Unfortunately, I can still reproduce with [723705e](https://github.com/James-Yu/LaTeX-Workshop/commit/723705e544a2c59b2ba6aafdf110db27faf51d39) . You may try like this: click the psf from the sidebar, then right click to delete the file. The opened editor will close despite the config is set to false.

The settings description states "Note that deleting from within the application will always close the editor", however, this is not the case in the linked issues, since those are deleted by external commands (e.g. latexmk). Deleting the pdf file from the commandline using `rm` does not trigger the editor to close for me.

**Reply by ManuelHentschel** [reply]
> Also, I’m against using two different approaches for viewing from the extension or clicking on the sidebar. This is 1. creating ambiguity if multiple tabs are opened from these two approaches, 2. effectively letting the extension to maintain codes solely for pdf viewing, which is not the objective of the extension, 3. hard to say whether the html listener to viewer events should be enabled or not on sidebar-clicked viewer.
> 
> May I bring up the question again, that is, what is missing from the current approach?

To me, the current approach feels somewhat clunky when working in a project that contains multiple pdf files. Even with the "singleton" option, opening pdfs behaves very different from opening other files. E.g., normal files can be opened only once per editor group, revealing the currently open editor, when the file is clicked in the sidebar, but it is possible to open the file multiple times in different tab groups. Trying to achieve this with the current approach seems to be impossible and leads to some buggy behavior, e.g. as follows:
- set `"latex-workshop.view.pdf.viewer": "singleton"`
- open a pdf in the left of two editor columns
- focus the right column and click on the same pdf in the explorer sidebar
- an empty web view will appear that disappears only after focussing something else

I agree that viewing pdfs is not the main point of this extension, but it is still a great feature of it, and is arguably part of the process of editing them. I really like the flexibility of viewing pdfs just like any other files (getting the behavior I expect from other editors/viewers) but then using synctex, automatic updates etc. Especially in large projects with multiple pdf files (e.g. different versions of the same document and lots of pdf-plots) this is really nice to have.

**Reply by James-Yu** [reply]
If I understand correctly, that is a possible bug in the singleton mode, which may have a fix. I’d also love to know other differences and contribute to fix them. In the meantime, there were numerous issues raised when we last time changed to custom editor. It is making me quite worried to change as the current solution works smoothly.

**Reply by ManuelHentschel** [reply]
Other differences include:
- Preview editors are not supported. Hence, quickly looking through a collection of PDFs (e.g. plots) leads to all of them staying open, instead of the default reuse behavior of normal editors
- The webview opened by vscode opens and closes immediately again, when clicking on a PDF that's already opened (and in some other scenarios, e.g. when multiple editors are opened at the same time)
- PDF editors don't use the correct icon
- Recent files (`ctrl+p`) shows each pdf twice while it's open
- #3410 seems to be fixed with the customEditor API

Each of these issues on its own might not be a big deal, but together make using the extension feel somewhat clunky, especially considering how straightforward the fix appears to be. However, I get that you want to be more cautious about changing stuff like this, and I would be happy to help maintain this "feature" and address any known bugs. Regarding the issues raised before, are there more than #3642, #3659? I tried finding related ones but did not find any in recent issues/discussions.

**Reply by James-Yu** [reply]
I will have a test shortly and come back to you soon (TM)

**Reply by James-Yu** [reply]
Before the test, any idea on how to preserve the original ‘tab’ behavior? I.e. multiple tabs for a same file.

**Reply by ManuelHentschel** [reply]
> Before the test, any idea on how to preserve the original ‘tab’ behavior? I.e. multiple tabs for a same file.

Currently, this does not seem to be possible (in a clean fashion). There are many requests for this feature at https://github.com/microsoft/vscode/issues/41289, so maybe it will be implemented at some point.

**Reply by James-Yu** [reply]
I confirm that https://github.com/James-Yu/LaTeX-Workshop/issues/3642, https://github.com/James-Yu/LaTeX-Workshop/discussions/3659 are fixed with the latest vscode + 723705e5 . The remaining issue is the original `tab` behavior, which I'd prefer maintaining.

As an alternative, I'd like to see the backend for `singleton` mode switched to customeditor, as the mode is still relatively new. I'd love to review it if you may have a PR on this. @ManuelHentschel 

---

edit: If customeditor is to incorporated, the new codes shall be as isolated as possible.

**Reply by ManuelHentschel** [reply]
Sounds good, I'll try to get a PR done by the end of the week!
